core/iwasm/common/wasm_shared_memory.c

```diff
@@ -417,8 +417,14 @@ wasm_runtime_atomic_wait(WASMModuleInstanceCommon *module, void *address,
 
     os_mutex_lock(&wait_info->wait_list_lock);
 
-    if ((!wait64 && *(uint32 *)address != (uint32)expect)
-        || (wait64 && *(uint64 *)address != expect)) {
+    WASMSharedMemNode *node =
+        search_module((WASMModuleCommon *)module_inst->module);
+    os_mutex_lock(&node->shared_mem_lock);
+    bool no_wait = (!wait64 && *(uint32 *)address != (uint32)expect)
+                   || (wait64 && *(uint64 *)address != expect);
+    os_mutex_unlock(&node->shared_mem_lock);
```

Comment:
The whole operation is already locked by wait_info->wait_list_lock, shall we add the lock node->shared_mem_lock again?

Reply:
For other atomic operations (for instance the atomic store), we acquired a lock before accessing the memory `maddr`
https://github.com/bytecodealliance/wasm-micro-runtime/blob/739acfc9082e63967ee91fca85dc01ab1630a5cc/core/iwasm/interpreter/wasm_interp_classic.c#L3557
But here (atomic wait)
https://github.com/bytecodealliance/wasm-micro-runtime/blob/739acfc9082e63967ee91fca85dc01ab1630a5cc/core/iwasm/interpreter/wasm_interp_classic.c#L3420
we don't do that